mitiq/zne/scaling/folding.py

```diff
@@ -421,6 +289,262 @@ def fold_all(
+            types which have a "moment" or "time" structure. Default is True.
+        return_mitiq (bool): If True, returns a Mitiq circuit instead of
+            the input circuit type (if different). Default is False.
+
+    Returns:
+        folded: the folded quantum circuit as a QPROGRAM.
+    """
+    _check_foldable(circuit)
+
+    if not (scale_factor >= 1):
+        raise ValueError("The scale factor must be a real number >= 1.")
+
+    folded = deepcopy(circuit)
+    measurements = _pop_measurements(folded)
+    base_circuit = deepcopy(folded)
+
+    # Determine the number of global folds and the final fractional scale
+    num_global_folds, fraction_scale = divmod(scale_factor - 1, 2)
+    # Do the global folds
+    for _ in range(int(num_global_folds)):
+        folded += Circuit(inverse(base_circuit), base_circuit)
+
+    # Fold remaining gates until the scale is reached
+    operations = list(base_circuit.all_operations())
+    num_to_fold = int(round(fraction_scale * len(operations) / 2))
+
+    if num_to_fold > 0:
+        folded += Circuit(
+            [inverse(operations[-num_to_fold:])], [operations[-num_to_fold:]]
+        )
+
+    _append_measurements(folded, measurements)
+    if not (kwargs.get("squash_moments") is False):
+        folded = _squash_moments(folded)
+    return folded
+
+
+def _create_weight_mask(
+    circuit: Circuit, fidelities: Optional[Dict[str, float]],
+) -> List[float]:
+    """Returns a list of weights associated to each gate if the input
+    circuit. Measurement gates are ignored.
+
+    Args:
+        circuit: The circuit from which a weight mask is created.
+        fidelities: The dictionary of gate fidelities.
+            If None, default fidelities will be used. See the
+            docstring of local folding function for mode details.
+
+    Returns: The list of weights associated to all the gates.
+    """
+    if fidelities and not all(0.0 < f <= 1.0 for f in fidelities.values()):
+        raise ValueError("Fidelities should be in the interval (0, 1].")
+    weights = None
+    if fidelities:
+        # Round to avoid ugly numbers like 0.09999999999999998 instead of 0.1
+        # https://docs.python.org/2/tutorial/floatingpoint.html
+        weights = {k: round(1.0 - f, 12) for k, f in fidelities.items()}
+
+    # Build mask with weights of each gate
+    return [
+        _get_weight_for_gate(weights, op)
+        for op in circuit.all_operations()
+        if not _is_measurement(op)
+    ]
+
+
+def _create_fold_mask(
+    weight_mask: List[float],
+    scale_factor: float,
+    folding_method: str,
+    seed: Optional[int] = None,
+) -> List[float]:
+    r"""Returns a list of integers determining how many times each gate a
+    circuit should be folded to realize the desired input scale_factor.
+
+    More precicely, the j_th element of the returned list is associated
+    to the j_th gate G_j of a circuit that we want to scale and determines
```

Comment:
"jth gate of a circuit" could be stated more clearly - a circuit is a 2d object. (What is the jth element of a matrix?)

Reply:
I clarified this in the docstring(s) of all the 3 functions related to masks.
This is the addition:
 ```
    The gate ordering is equal to the one used in the `all_operations()`
    method of the :class:`cirq.Circuit` class: gates from earlier moments
    come first and gates within the same moment follow the order in which
    they were given to the moment's constructor.
```

This comment is similar to the docstring of `all_operations`:

https://github.com/quantumlib/Cirq/blob/0ab45cf24e53a0de3b60404ce5f713dc92491cc3/cirq/circuits/circuit.py#L868-L875